File: C/schritt.c
#include "schritt.h"
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

SchrittListe lade_schritte_aus_file(const char *filename) {
  SchrittListe liste;
  liste.data = NULL;
  liste.count = 0;

  FILE *f = fopen(filename, "r");
  if (!f) {
    perror("Fehler beim Öffnen der Datei");
    return liste;
  }

  char line[256];
  while (fgets(line, sizeof(line), f)) {
    line[strcspn(line, "\r\n")] = 0; // Whitespace entfernen (wie strip())

    if (strlen(line) == 0) // leere Zeile überspringen
      continue;

    char *words[8]; // Wörter splitten
    int i = 0;
    char *token = strtok(line, " ");
    while (token && i < 8) {
      words[i++] = token;
      token = strtok(NULL, " ");
    }

    if (i < 8)
      continue; // fehlerhafte Zeile

    liste.data = realloc(liste.data, (liste.count + 1) * sizeof(Schritt));
    Schritt *s = &liste.data[liste.count];

    snprintf(s->ziel_typ, sizeof(s->ziel_typ), "%s", words[0]);
    snprintf(s->ziel_bezeichnung, sizeof(s->ziel_bezeichnung), "%s", words[1]);
    s->schritt_nr = atof(words[2]);
    snprintf(s->quell_typ, sizeof(s->quell_typ), "%s", words[3]);
    snprintf(s->quell_bezeichnung, sizeof(s->quell_bezeichnung), "%s", words[4]);
    s->menge = atof(words[5]);
    snprintf(s->taetigkeit, sizeof(s->taetigkeit), "%s", words[6]);
    s->zeit = atof(words[7]);
    liste.count++;
  }

  fclose(f);
  return liste;
}

int main_schritt(int argc, char *argv[]) {
  const char *filename = "schritt.dat";
  SchrittListe schritte = lade_schritte_aus_file(filename);

  for (int i = 0; i < schritte.count; i++) {
    printf(
        "ziel_typ: %s\nziel_bezeichnung: %s\nschritt_nr: %d\nquell_typ: "

        "%s\nquell_bezeichnung: %s\nmenge: %f\ntaetigkeit: %s\nzeit: %.2f\n\n",
        schritte.data[i].ziel_typ, schritte.data[i].ziel_bezeichnung,
        schritte.data[i].schritt_nr, schritte.data[i].quell_typ,
        schritte.data[i].quell_bezeichnung, schritte.data[i].menge,
        schritte.data[i].taetigkeit, schritte.data[i].zeit);
  }

  free(schritte.data);

  return EXIT_SUCCESS;
}

File: Python/schritt.py
from typing import List


def lade_schritte_aus_tsv_file(tsv_file: str) -> List["Schritt"]:
    schritte: List["Schritt"] = []
    with open(tsv_file, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            words = line.split()

            if line:  # leere Zeilen überspringen
                schritt = Schritt(
                    words[0],
                    words[1],
                    words[2],
                    words[3],
                    words[4],
                    words[5],
                    words[6],
                    words[7],
                )
                schritte.append(schritt)
    return schritte


class Schritt:
    def __init__(
        self,
        ziel_typ,
        ziel_bez,
        schritt_nr,
        quell_typ,
        quell_bez,
        menge,
        taetigkeit,
        zeit,
    ) -> None:
        self.ziel_typ = ziel_typ
        self.ziel_bez = ziel_bez
        self.schritt_nr = schritt_nr
        self.quell_typ = quell_typ
        self.quell_bez = quell_bez
        self.menge = menge
        self.taetigkeit = taetigkeit
        self.zeit = zeit
        self.benoetigtes_teil = self.erstelle_benoetigtes_teil()

    def erstelle_benoetigtes_teil(self) -> List[str]:
        benoetigtes_teil: List[str] = [self.quell_typ, self.quell_bez]
        return benoetigtes_teil

    def __repr__(self):
        return f""" 
    ziel_typ: {self.ziel_typ}
    ziel_bez: {self.ziel_bez} 
    schritt_nr: {self.schritt_nr}
    quell_typ: {self.quell_typ}
    quell_bez: {self.quell_bez} 
    menge: {self.menge}
    taetigkeit: {self.taetigkeit}
    zeit: {self.zeit}
    """


def main():
    schritte = lade_schritte_aus_tsv_file("schritt.dat")
    for schritt in schritte:
        print(schritt)


if __name__ == "__main__":
    main()
